src/main/java/com/avispl/symphony/dal/communicator/middleatlantic/MiddleAtlanticPowerUnitCommunicator.java

```diff
@@ -38,57 +41,63 @@ public MiddleAtlanticPowerUnitCommunicator() {
         // it can be changed back by configuration
         setTrustAllCertificates(true);
-        localStatistics = new ExtendedStatistics();
-        localStatistics.setControl(new HashMap<>());
-        localStatistics.setStatistics(new HashMap<>());
     }
 
     @Override
     protected void authenticate() throws Exception {
         // nothing to do here, authentication is done in individual requests
     }
 
+    /**
+     * Getting statistics of the device.
+     * Need to make sure that none of control operations are running when trying to fetch the statistics
+     */
     @Override
     public List<Statistics> getMultipleStatistics() throws Exception {
+        ExtendedStatistics extendedStatistics = new ExtendedStatistics();
         // This is to make sure if the statistics is being fetched before/after any set of control operations
-        if(!controlsRunning) {
-            try {
-                readWriteLock.writeLock().lock();
-                ExtendedStatistics extendedStatistics = new ExtendedStatistics();
-                Map<String, String> statistics = Collections.synchronizedMap(new LinkedHashMap<>());
-                Map<String, String> control = new ConcurrentHashMap<>();
-
-                ExecutorService executor = Executors.newCachedThreadPool();
-                runAsync(() -> fillInStatistics(statistics, "Inlet Active Power", "/inlet/0/activePower", GET_READING), executor)
-                        .thenRunAsync(() -> fillInStatistics(statistics, "Inlet RMS Current", "/inlet/0/current", GET_READING), executor)
-                        .thenRunAsync(() -> fillInStatistics(statistics, "Inlet Line Frequency", "/inlet/0/lineFrequency", GET_READING), executor)
-                        .thenRunAsync(() -> fillInStatistics(statistics, "Inlet RMS Voltage", "/inlet/0/voltage", GET_READING), executor)
-                        .thenApplyAsync(ignore -> getOutletsCount(), executor)
-                        .thenAcceptAsync(count -> IntStream.range(0, count)
-                                .parallel()
-                                .mapToObj(num -> runAsync(() -> fillInOutletState(statistics, control, num), executor)
-                                        .thenRunAsync(() -> fillInStatistics(statistics, getOutletRMSName(num),
-                                                "/outlet/" + num + "/current", GET_READING), executor)
-                                )
-                                .forEach(CompletableFuture::join), executor)
-                        .get(30, TimeUnit.SECONDS);
-
-                executor.shutdownNow();
-
-                extendedStatistics.setStatistics(statistics);
-                extendedStatistics.setControl(control);
-                localStatistics = extendedStatistics;
-            } finally {
-                readWriteLock.writeLock().unlock();
-            }
+        if(controlsRunning && localStatistics != null){
```

Comment:
It is not thread safe anymore - now the localStatistics is accessed outside of the lock. It can be set to null at any time, of modified by control thread.